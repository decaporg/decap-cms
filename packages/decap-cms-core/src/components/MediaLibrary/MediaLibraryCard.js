import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from '@emotion/styled';
import { colors, borders, lengths, shadows, effects } from 'decap-cms-ui-default';

const IMAGE_HEIGHT = 160;

const Card = styled.div`
  width: ${props => props.width};
  height: ${props => props.height};
  margin: ${props => props.margin};
  border: ${borders.textField};
  border-color: ${props => props.isSelected && colors.active};
  border-radius: ${lengths.borderRadius};
  cursor: pointer;
  overflow: hidden;
  background-color: ${props => props.isPrivate && colors.textFieldBorder};

  &:focus {
    outline: none;
  }
`;

const CardImageWrapper = styled.div`
  height: ${IMAGE_HEIGHT + 2}px;
  ${effects.checkerboard};
  ${shadows.inset};
  border-bottom: solid ${lengths.borderWidth} ${colors.textFieldBorder};
  position: relative;
`;

const CardImage = styled.img`
  width: 100%;
  height: ${IMAGE_HEIGHT}px;
  object-fit: contain;
  border-radius: 2px 2px 0 0;
`;

const CardFileIcon = styled.div`
  width: 100%;
  height: 160px;
  object-fit: cover;
  border-radius: 2px 2px 0 0;
  padding: 1em;
  font-size: 3em;
`;

const CardText = styled.p`
  color: ${colors.text};
  padding: 8px;
  margin-top: 20px;
  overflow-wrap: break-word;
  line-height: 1.3 !important;
`;

const DraftText = styled.p`
  color: ${colors.mediaDraftText};
  background-color: ${colors.mediaDraftBackground};
  position: absolute;
  padding: 8px;
  border-radius: ${lengths.borderRadius} 0 ${lengths.borderRadius} 0;
`;

class MediaLibraryCard extends React.Component {
  render() {
    const {
      isSelected,
      displayURL,
      text,
      onClick,
      draftText,
      width,
      height,
      margin,
      isPrivate,
      type,
      isViewableImage,
      isDraft,
    } = this.props;
    const url = displayURL.get('url');
    return (
      <Card
        isSelected={isSelected}
        onClick={onClick}
        width={width}
        height={height}
        margin={margin}
        tabIndex="-1"
        isPrivate={isPrivate}
      >
        <CardImageWrapper>
          {isDraft ? <DraftText data-testid="draft-text">{draftText}</DraftText> : null}
          {url && isViewableImage ? (
            <CardImage loading="lazy" src={url} />
          ) : (
            <CardFileIcon data-testid="card-file-icon">{type}</CardFileIcon>
          )}
        </CardImageWrapper>
        <CardText>{text}</CardText>
      </Card>
    );
  }
  componentDidMount() {
    const { displayURL, loadDisplayURL } = this.props;
    if (!displayURL.get('url')) {
      loadDisplayURL();
    }
  }
}

MediaLibraryCard.propTypes = {
  isSelected: PropTypes.bool,
  displayURL: ImmutablePropTypes.map.isRequired,
  text: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  draftText: PropTypes.string.isRequired,
  width: PropTypes.string.isRequired,
  height: PropTypes.string.isRequired,
  margin: PropTypes.string.isRequired,
  isPrivate: PropTypes.bool,
  type: PropTypes.string,
  isViewableImage: PropTypes.bool.isRequired,
  loadDisplayURL: PropTypes.func.isRequired,
  isDraft: PropTypes.bool,
};

export default MediaLibraryCard;
