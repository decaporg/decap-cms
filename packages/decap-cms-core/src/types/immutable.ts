import { Map as ImmutableMap, List } from 'immutable';

export function isImmutableMap(value: unknown): value is ImmutableMap<string, unknown> {
  return ImmutableMap.isMap(value);
}

export function isImmutableList(value: unknown): value is List<unknown> {
  return List.isList(value);
}

export interface StaticallyTypedRecord<T> {
  get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K];
  set<K extends keyof T, V extends T[K]>(key: K, value: V): StaticallyTypedRecord<T> & T;
  has<K extends keyof T>(key: K): boolean;
  delete<K extends keyof T>(key: K): StaticallyTypedRecord<T>;
  getIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    defaultValue?: V,
  ): T[K1][K2];
  getIn<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    V extends T[K1][K2][K3],
  >(
    keys: [K1, K2, K3],
    defaultValue?: V,
  ): T[K1][K2][K3];
  getIn(keys: string[]): unknown;
  setIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    value: V,
  ): StaticallyTypedRecord<T>;
  setIn(keys: string[], value: unknown): StaticallyTypedRecord<T> & T;
  toJS(): T;
  isEmpty(): boolean;
  some<K extends keyof T>(predicate: (value: T[K], key: K, iter: this) => boolean): boolean;
  mapKeys<K extends keyof T, V>(mapFunc: (key: K, value: StaticallyTypedRecord<T>) => V): V[];
  find<K extends keyof T>(findFunc: (value: T[K]) => boolean): T[K];
  filter<K extends keyof T>(
    predicate: (value: T[K], key: K, iter: this) => boolean,
  ): StaticallyTypedRecord<T>;
  valueSeq<K extends keyof T>(): T[K][] & { toArray: () => T[K][] };
  map<K extends keyof T, V>(
    mapFunc: (value: T[K]) => V,
  ): StaticallyTypedRecord<{ [key: string]: V }>;
  keySeq<K extends keyof T>(): { toArray: () => K[] };
  withMutations(mutator: (mutable: StaticallyTypedRecord<T>) => unknown): StaticallyTypedRecord<T>;
}
