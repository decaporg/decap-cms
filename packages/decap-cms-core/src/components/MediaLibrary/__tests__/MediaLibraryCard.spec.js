import React from 'react';
import { Map } from 'immutable';
import { render } from '@testing-library/react';

import MediaLibraryCard from '../MediaLibraryCard';

describe('MediaLibraryCard', () => {
  const props = {
    displayURL: Map({ url: 'url' }),
    text: 'image.png',
    onClick: jest.fn(),
    draftText: 'Draft',
    width: '100px',
    height: '240px',
    margin: '10px',
    isViewableImage: true,
    loadDisplayURL: jest.fn(),
  };

  it('should match snapshot for non draft image', () => {
    const { asFragment, queryByTestId } = render(<MediaLibraryCard {...props} />);

    expect(queryByTestId('draft-text')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot for draft image', () => {
    const { asFragment, getByTestId } = render(<MediaLibraryCard {...props} isDraft={true} />);
    expect(getByTestId('draft-text')).toHaveTextContent('Draft');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot for non viewable image', () => {
    const { asFragment, getByTestId } = render(
      <MediaLibraryCard {...props} isViewableImage={false} type="Not Viewable" />,
    );
    expect(getByTestId('card-file-icon')).toHaveTextContent('Not Viewable');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call loadDisplayURL on mount when url is empty', () => {
    const loadDisplayURL = jest.fn();
    render(
      <MediaLibraryCard {...props} loadDisplayURL={loadDisplayURL} displayURL={Map({ url: '' })} />,
    );

    expect(loadDisplayURL).toHaveBeenCalledTimes(1);
  });
});
