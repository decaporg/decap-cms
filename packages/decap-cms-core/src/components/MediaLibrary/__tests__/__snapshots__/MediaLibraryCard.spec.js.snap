// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaLibraryCard should match snapshot for draft image 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100px;
  height: 240px;
  margin: 10px;
  border: solid 2px #dfdfe3;
  border-radius: 5px;
  cursor: pointer;
  overflow: hidden;
}

.emotion-0:focus {
  outline: none;
}

.emotion-2 {
  height: 162px;
  background-color: #f2f2f2;
  -webkit-background-size: 16px 16px;
  background-size: 16px 16px;
  -webkit-background-position: 0 0,8px 8px;
  background-position: 0 0,8px 8px;
  background-image: linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      ),linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      );
  box-shadow: inset 0 0 4px rgba(68, 74, 87, 0.3);
  border-bottom: solid 2px #dfdfe3;
  position: relative;
}

.emotion-4 {
  color: #70399f;
  background-color: #f6d8ff;
  position: absolute;
  padding: 8px;
  border-radius: 5px 0 5px 0;
}

.emotion-6 {
  width: 100%;
  height: 160px;
  object-fit: contain;
  border-radius: 2px 2px 0 0;
}

.emotion-8 {
  color: #798291;
  padding: 8px;
  margin-top: 20px;
  overflow-wrap: break-word;
  line-height: 1.3!important;
}

<div
    class="emotion-0 emotion-1"
    height="240px"
    tabindex="-1"
    width="100px"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <p
        class="emotion-4 emotion-5"
        data-testid="draft-text"
      >
        Draft
      </p>
      <img
        class="emotion-6 emotion-7"
        loading="lazy"
        src="url"
      />
    </div>
    <p
      class="emotion-8 emotion-9"
    >
      image.png
    </p>
  </div>
</DocumentFragment>
`;

exports[`MediaLibraryCard should match snapshot for non draft image 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100px;
  height: 240px;
  margin: 10px;
  border: solid 2px #dfdfe3;
  border-radius: 5px;
  cursor: pointer;
  overflow: hidden;
}

.emotion-0:focus {
  outline: none;
}

.emotion-2 {
  height: 162px;
  background-color: #f2f2f2;
  -webkit-background-size: 16px 16px;
  background-size: 16px 16px;
  -webkit-background-position: 0 0,8px 8px;
  background-position: 0 0,8px 8px;
  background-image: linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      ),linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      );
  box-shadow: inset 0 0 4px rgba(68, 74, 87, 0.3);
  border-bottom: solid 2px #dfdfe3;
  position: relative;
}

.emotion-4 {
  width: 100%;
  height: 160px;
  object-fit: contain;
  border-radius: 2px 2px 0 0;
}

.emotion-6 {
  color: #798291;
  padding: 8px;
  margin-top: 20px;
  overflow-wrap: break-word;
  line-height: 1.3!important;
}

<div
    class="emotion-0 emotion-1"
    height="240px"
    tabindex="-1"
    width="100px"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <img
        class="emotion-4 emotion-5"
        loading="lazy"
        src="url"
      />
    </div>
    <p
      class="emotion-6 emotion-7"
    >
      image.png
    </p>
  </div>
</DocumentFragment>
`;

exports[`MediaLibraryCard should match snapshot for non viewable image 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100px;
  height: 240px;
  margin: 10px;
  border: solid 2px #dfdfe3;
  border-radius: 5px;
  cursor: pointer;
  overflow: hidden;
}

.emotion-0:focus {
  outline: none;
}

.emotion-2 {
  height: 162px;
  background-color: #f2f2f2;
  -webkit-background-size: 16px 16px;
  background-size: 16px 16px;
  -webkit-background-position: 0 0,8px 8px;
  background-position: 0 0,8px 8px;
  background-image: linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      ),linear-gradient(
        45deg,
        #e6e6e6 25%,
        transparent 25%,
        transparent 75%,
        #e6e6e6 75%,
        #e6e6e6
      );
  box-shadow: inset 0 0 4px rgba(68, 74, 87, 0.3);
  border-bottom: solid 2px #dfdfe3;
  position: relative;
}

.emotion-4 {
  width: 100%;
  height: 160px;
  object-fit: cover;
  border-radius: 2px 2px 0 0;
  padding: 1em;
  font-size: 3em;
}

.emotion-6 {
  color: #798291;
  padding: 8px;
  margin-top: 20px;
  overflow-wrap: break-word;
  line-height: 1.3!important;
}

<div
    class="emotion-0 emotion-1"
    height="240px"
    tabindex="-1"
    width="100px"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4 emotion-5"
        data-testid="card-file-icon"
      >
        Not Viewable
      </div>
    </div>
    <p
      class="emotion-6 emotion-7"
    >
      image.png
    </p>
  </div>
</DocumentFragment>
`;
